# 1.) Import the Credit Card Fraud Data From CCLE

In [ ]:
import pandas as pd
from google.colab import drive
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
drive.mount('/content/gdrive/', force_remount = True)

Mounted at /content/gdrive/


In [ ]:
df = pd.read_csv("/content/gdrive/MyDrive/Econ 441B/fraudTest.csv")

In [ ]:
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

# 2.) Select four columns to use as features (one just be trans_date_trans)

In [ ]:
df_select = df[["trans_date_trans_time", "category", "amt", "city_pop", "is_fraud"]]

In [ ]:
df_select.columns

Index(['trans_date_trans_time', 'category', 'amt', 'city_pop', 'is_fraud'], dtype='object')

In [ ]:
df_select = pd.get_dummies(df_select, columns=["category"])

In [ ]:
df_select.head()

trans_date_trans_time    amt  city_pop  is_fraud  category_entertainment  \
0   2020-06-21 12:14:25   2.86    333497         0                       0   
1   2020-06-21 12:14:33  29.84       302         0                       0   
2   2020-06-21 12:14:53  41.28     34496         0                       0   
3   2020-06-21 12:15:15  60.05     54767         0                       0   
4   2020-06-21 12:15:17   3.19      1126         0                       0   

   category_food_dining  category_gas_transport  category_grocery_net  \
0                     0                       0                     0   
1                     0                       0                     0   
2                     0                       0                     0   
3                     0                       0                     0   
4                     0                       0                     0   

   category_grocery_pos  category_health_fitness  category_home  \
0                     0                        0              0   
1                     0                        0              0   
2                     0                        1              0   
3                     0                        0              0   
4                     0                        0              0   

   category_kids_pets  category_misc_net  category_misc_pos  \
0                   0                  0                  0   
1                   0                  0                  0   
2                   0                  0                  0   
3                   0                  0                  1   
4                   0                  0                  0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                       1                      0                      0   
1                       1                      0                      0   
2                       0                      0                      0   
3                       0                      0                      0   
4                       0                      0                      0   

   category_travel  
0                0  
1                0  
2                0  
3                0  
4                1

# 3.) Create a unique variable out of trans_date.

In [ ]:
type(df_select["trans_date_trans_time"][0])

str

In [ ]:
df_select["trans_date_trans_time"] = pd.to_datetime(df_select["trans_date_trans_time"])
#dir(df_select["trans_date_trans_time"])

In [ ]:
df["trans_date_trans_time"].head()

0    2020-06-21 12:14:25
1    2020-06-21 12:14:33
2    2020-06-21 12:14:53
3    2020-06-21 12:15:15
4    2020-06-21 12:15:17
Name: trans_date_trans_time, dtype: object

In [ ]:
df_select["time"] = [i.hour for i in df_select["trans_date_trans_time"]]

In [ ]:
x = df_select.drop(["trans_date_trans_time", "is_fraud"], axis=1)
y = df["is_fraud"]

In [ ]:
x.head()

amt  city_pop  category_entertainment  category_food_dining  \
0   2.86    333497                       0                     0   
1  29.84       302                       0                     0   
2  41.28     34496                       0                     0   
3  60.05     54767                       0                     0   
4   3.19      1126                       0                     0   

   category_gas_transport  category_grocery_net  category_grocery_pos  \
0                       0                     0                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   

   category_health_fitness  category_home  category_kids_pets  \
0                        0              0                   0   
1                        0              0                   0   
2                        1              0                   0   
3                        0              0                   0   
4                        0              0                   0   

   category_misc_net  category_misc_pos  category_personal_care  \
0                  0                  0                       1   
1                  0                  0                       1   
2                  0                  0                       0   
3                  0                  1                       0   
4                  0                  0                       0   

   category_shopping_net  category_shopping_pos  category_travel  time  
0                      0                      0                0    12  
1                      0                      0                0    12  
2                      0                      0                0    12  
3                      0                      0                0    12  
4                      0                      0                1    12

In [ ]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: is_fraud, dtype: int64

# 4.) Oversample the data (this will be your training data).

In [ ]:
cut = int((len(df_select) * .8) // 1)

x_train = x[:cut]
x_test = x[cut:]
y_train = y[:cut]
y_test = y[cut:]

In [ ]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import RandomOverSampler

In [ ]:
# Randomly over sample
ros = RandomOverSampler(random_state=42)
x_train_ros, y_train_ros= ros.fit_resample(x_train, y_train)
# Check the number of records after over sampling
print(sorted(Counter(y_train_ros).items()))

[(0, 442614), (1, 442614)]


# 5.) Train a Logistic regression.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_reg = LogisticRegression().fit(x_train_ros, y_train_ros)

# 6.) The company you are working for wants to target at a False Positive rate of 5% what threshold should you use? (Use oversampled data)

In [ ]:
# use chatGPT
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_curve

# predict the probabilities of the positive class for the test data
y_prob = log_reg.predict_proba(x_test)[:,1]

# calculate the false positive rate and true positive rate
fp, tp, thresholds = roc_curve(y_test, y_prob)

# find the threshold that corresponds to a false positive rate of 5%
threshold = thresholds[np.argmin(np.abs(fp - 0.05))]

# use this threshold to make predictions on the test data
y_pred = (y_prob > threshold).astype(int)

# create the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
conf_matrix

array([[105152,   5808],
       [    62,    122]])

In [ ]:
threshold

0.6278003605398791

To target at a False Positive rate of 5%, we should use a threshold of 0.6278.

# 7.) If the company makes .02*amt on True transactions and loses -amt on False (Use original data)

In [ ]:
df_select['predict']=log_reg.predict(x)
df_select.head()

trans_date_trans_time    amt  city_pop  is_fraud  category_entertainment  \
0   2020-06-21 12:14:25   2.86    333497         0                       0   
1   2020-06-21 12:14:33  29.84       302         0                       0   
2   2020-06-21 12:14:53  41.28     34496         0                       0   
3   2020-06-21 12:15:15  60.05     54767         0                       0   
4   2020-06-21 12:15:17   3.19      1126         0                       0   

   category_food_dining  category_gas_transport  category_grocery_net  \
0                     0                       0                     0   
1                     0                       0                     0   
2                     0                       0                     0   
3                     0                       0                     0   
4                     0                       0                     0   

   category_grocery_pos  category_health_fitness  category_home  \
0                     0                        0              0   
1                     0                        0              0   
2                     0                        1              0   
3                     0                        0              0   
4                     0                        0              0   

   category_kids_pets  category_misc_net  category_misc_pos  \
0                   0                  0                  0   
1                   0                  0                  0   
2                   0                  0                  0   
3                   0                  0                  1   
4                   0                  0                  0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                       1                      0                      0   
1                       1                      0                      0   
2                       0                      0                      0   
3                       0                      0                      0   
4                       0                      0                      0   

   category_travel  time  predict  
0                0    12        0  
1                0    12        0  
2                0    12        0  
3                0    12        0  
4                1    12        0

In [ ]:
# Define the profit using the prediction, actual, and amt
def profit(df_select):
  profit = 0
  for i in range(len(df_select)):
    if df_select['predict'][i]==1 and df_select['is_fraud'][i]==1:
      profit = profit + 0.02*df_select['amt'][i]
    elif df_select['predict'][i]==0 and df_select['is_fraud'][i]==1:
      profit = profit - df_select['amt'][i]
    else:
      profit = profit
  return profit

In [ ]:
profit(df_select)

-3889.746999999991

The profit if the company makes .02*amt on True transactions and loses -amt on False is -$3,889.75.

# 8.) Using Logistic Regression Lasso to inform you. Would you use the selected features in a trusted prediction model?

In [ ]:
reg = LogisticRegression(penalty = 'l1', solver='liblinear').fit(x_train, y_train)

In [ ]:
reg.coef_

array([[ 2.25749368e-03, -7.18346796e-07, -2.72166992e+00,
        -2.85446373e+00, -8.51697331e-01, -9.54716331e-01,
         2.06981084e-01, -2.92086924e+00, -3.14774416e+00,
        -2.98487094e+00, -2.42805784e-01, -2.04463448e+00,
        -2.64247223e+00, -8.04304175e-01, -2.04204497e+00,
        -6.80175442e+00,  1.22316449e-01]])

Yes, because most of the coefficients have a value other than zero, I would use the selected features in a prediction model.